mlir/lib/Catalyst/Transforms/DetectQNodes.cpp

```diff
@@ -0,0 +1,1272 @@
+    return std::tuple<std::vector<Value>, std::vector<Value>>(collectedTokens, collectedValues);
+}
+
+void insertCallToMlirAsyncRuntimeErrorFunction(Value value, LLVM::LLVMFuncOp fnDecl,
+                                               Block *failBlock, PatternRewriter &rewriter)
+{
+    PatternRewriter::InsertionGuard insertGuard(rewriter);
+    rewriter.setInsertionPointToEnd(failBlock);
+    SmallVector<Value> operands = {value};
+    rewriter.create<LLVM::CallOp>(fnDecl.getLoc(), fnDecl, operands);
+}
+
+void insertErrorCalls(std::vector<Value> tokens, std::vector<Value> values, Block *failBlock,
+                      PatternRewriter &rewriter)
+{
+    // At the fail block, it is guaranteed that all runtime values are available
+    // but they
+    PatternRewriter::InsertionGuard insertGuard(rewriter);
+    rewriter.setInsertionPointToEnd(failBlock);
+    auto landingPad = failBlock->begin();
+    auto moduleOp = landingPad->getParentOfType<ModuleOp>();
+
+    LLVM::LLVMFuncOp setTokenError = lookupOrCreateMlirAsyncRuntimeSetTokenError(moduleOp);
+    LLVM::LLVMFuncOp setValueError = lookupOrCreateMlirAsyncRuntimeSetValueError(moduleOp);
+    for (auto token : tokens) {
+        insertCallToMlirAsyncRuntimeErrorFunction(token, setTokenError, failBlock, rewriter);
+    }
+    for (auto value : values) {
+        insertCallToMlirAsyncRuntimeErrorFunction(value, setValueError, failBlock, rewriter);
+    }
+}
+
+void insertBranchFromFailToSuccessor(Block *fail, Block *success, PatternRewriter &rewriter)
+{
+    // The reason why we are unconditionally jumping from failure to success
+    // is because the failure is communicated through the state of the runtime tokens
+    // and values.
+    //
+    // The error will finally be managed by the main thread.
+    // But this thread is expected to end normally.
+    PatternRewriter::InsertionGuard insertGuard(rewriter);
+    rewriter.setInsertionPointToEnd(fail);
+
+    auto landingPad = fail->begin();
+    auto loc = landingPad->getLoc();
+
+    rewriter.create<LLVM::BrOp>(loc, success);
+}
+
+void cleanupPreHandleErrorAttr(LLVM::LLVMFuncOp funcOp, PatternRewriter &rewriter)
+{
+    rewriter.updateRootInPlace(funcOp, [&] { funcOp->removeAttr(preHandleErrorAttrValue); });
+}
+
+void scheduleAnalysisForErrorHandling(LLVM::LLVMFuncOp funcOp, PatternRewriter &rewriter)
+{
+    rewriter.updateRootInPlace(
+        funcOp, [&] { funcOp->setAttr(preHandleErrorAttrValue, rewriter.getUnitAttr()); });
+}
+
+void scheduleLivenessAnalysis(LLVM::CallOp callOp, PatternRewriter &rewriter)
+{
+    rewriter.updateRootInPlace(callOp, [&] { callOp->setAttr(sinkAttr, rewriter.getUnitAttr()); });
+}
+
+void scheduleCallToInvoke(LLVM::CallOp callOp, PatternRewriter &rewriter)
+{
+    rewriter.updateRootInPlace(
+        callOp, [&] { callOp->setAttr(scheduleInvokeAttr, rewriter.getUnitAttr()); });
+}
+
+bool isAsync(LLVM::LLVMFuncOp funcOp)
+{
+    if (!funcOp->hasAttr(passthroughAttr))
+        return false;
```

Comment:
consider this suggestion for the code consistency:
 ```suggestion
    if (!funcOp->hasAttr(passthroughAttr)) {
        return false;
    }
```